[PLATFORM-1146]: Write step by step documentation (#171)
primait/event_sourcing.rs: README.md

cargo make clippy
```

## Usage

> In this section is presented the integration of this library into your application, enabling you to kickstart 

Reviewer: ```suggestion
In this section we present how to integrate this library into your application, enabling you to kickstart 
```

Author: I like much more to leave it impersonal. What the others think?

> CQRS/Event sourcing implementation.

Reviewer: ```suggestion
your CQRS/Event sourcing implementation.
```

Reviewer 2: Either `your` or `an` makes the text flow better imo
```suggestion
an CQRS/Event sourcing implementation.
```

Author: Maye `a CQRS`?

Reviewer 2: Hm, yeah, I think you're right, but I'm not 100% sure

### Event sourcing

> Event sourcing is a software architectural pattern used to capture and store the changes or events that occur in an 
application's state over time. Instead of persisting the current state of an object or entity, event sourcing focuses 
on recording a sequence of events that have led to that state.

#### Aggregate

> To initiate the implementation of CQRS/Event sourcing in your codebase, the initial building block to create is an `Aggregate`.

Reviewer: ```suggestion
To start the implementation of CQRS/Event sourcing in your codebase, the initial building block you need to create is an `Aggregate`.
```

Reviewer 2: I think
```suggestion
To implement CQRS/Event sourcing in your codebase, you need to start by creating an `Aggregate`.
```
is a bit better than what Cristiano suggested?
We could also do
```suggestion
To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
```

Author: I think that i will go impersonal to maintain the overall style being that above we agreed on changing text to `This section presents the integration of the `

Reviewer 2: I think
> To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
is impersonal, but flows a bit better

> In the context of event sourcing, an aggregate is a fundamental concept that represents a cluster of domain objects 
and their state changes. It is one of the key building blocks used to model and manage data in an event-sourced system.

> An aggregate, therefore, is responsible for processing incoming commands and generating corresponding events that 
reflect the changes to its state. It encapsulates domain logic and rules related to specific operations on the data it 

> manages. Aggregates provide a clear boundary for consistency and concurrency control within the event-sourced system.

Reviewer: If this is a quote I would put a `source` if possible

Author: Mmmh no source right there.. Should i remove the quotation?

Reviewer 2: If it's not a quote I think having it in plain text is more sensible

    NotEnoughCopies,
}
```

And now let's put all together in the `Aggregate`, implementing `handle_command` and `apply_events` functions too.

```rust
...

impl Aggregate for Book {
    const NAME: &'static str = "book";
    type State = BookState;
    type Command = BookCommand;
    type Event = BookEvent;
    type Error = BookError;

    fn handle_command(state: &Self::State, command: Self::Command) -> Result<Vec<Self::Event>, Self::Error> {
        match command {
            BookCommand::Buy { num_of_copies } if state.leftover < num_of_copies => Err(BookError::NotEnoughCopies),
            BookCommand::Buy { num_of_copies } => Ok(vec![BookEvent::Bought { num_of_copies }]),
            BookCommand::Return { num_of_copies } => Ok(vec![BookEvent::Returned { num_of_copies }]),
        }
    }

    fn apply_event(state: Self::State, payload: Self::Event) -> Self::State {
        match payload {
            BookEvent::Bought { num_of_copies } => BookState { leftover: state.leftover - num_of_copies },
            BookEvent::Returned { num_of_copies } => BookState { leftover: state.leftover + num_of_copies },
        }
    }
}
```

#### Adding a persistence layer

> At the time being the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated

Reviewer: ```suggestion
For the time being the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated
```

or

```suggestion
Currently, the only existing persistence layer is the `PgStore`. Its role is to write all the events in a dedicated
```

table in `postgres`. The table name is built using `Aggregate::NAME` constant value concatenated with `_events`.

!Important: each `PgStore` is designed to be globally unique per Aggregate. If you require constructing it dynamically 
at runtime, remember to invoke `without_running_migrations` while building the store to avoid executing migrations.

> Note: In order to use a `PgStore` it's needed a fully implemented `Aggregate`.

Reviewer: ```suggestion
Note: In order to use a `PgStore` you need a fully implemented `Aggregate`.
```

write operations (the Command side).

#### The sync way

In the purist concept of CQRS, this approach might provoke debates. Nevertheless, the underlying intent is to provide 
timely feedback to users upon completion of operations, considering various scenarios.

For instance, it aims to eliminate the need for a hypothetical frontend to continually poll the backend for updates, by 
ensuring the read side is promptly updated. Additionally, it caters to scenarios where a sequence of operations forms a 
chain, and completing the entire task is crucial, as in the case of a SAGA pattern.

Overall, the primary goal is to enhance user experience and system efficiency by employing well-defined and specialized 
models for read and write operations, tailoring the architecture to specific use cases.

The two main pillars in the context of CQRS/Event sourcing are "Event Handlers" and "Transactional Event Handlers."

##### Event handlers

An `EventHandler`, by definition, operates on an eventually consistent basis and primarily serves to update the read side 
of the application. It is also commonly utilized to execute commands on other aggregates, including the aggregate to 
which it belongs.

> `EventHandler`s are infallible.

Reviewer: I would elaborate a bit on what infallible means in this context

#### Aggregate

> To start implementing of CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.

Reviewer 2: ```suggestion
To start implementing CQRS/Event sourcing in your codebase, first you need to create an `Aggregate`.
```suggestion
